pyccel/ast/core.py

```diff
@@ -2367,6 +2368,10 @@ def process_shape(self, shape):
     def name(self):
         return self._name
 
+    @property
+    def alloc_shape(self):
+        return self._alloc_shape
+
```

Comment:
It is the shape used for the allocation/declaration. If we take the following example:
```
import numpy as np
from pyccel.decorators import stack_array, types

@stack_array('a')
@types('int')
def f(n):
    a = np.ones(n)
    n = 10
    b = np.ones_like(a)
```
Then the shape of `a` needs to be saved as `(shape(a,0),)` as the expression describing it's shape is not constant. However as `a` is a stack array it's declaration needs to contain information about `n`

Similarly for the following example:
```
import numpy as np

def g():
    import numpy as np
    c = np.zeros(np.random.randint(23))
    d = np.zeros_like(c)
```
The allocation shape of `c` is the randint expression however the shape of `c` which is used whenever we manipulate it is `(shape(c,0),)`

In the second case the allocation shape of `c` is not actually necessary as the `Zeros` object has shape `(np.random.randint(23)` which is used for the allocation.

Maybe I should call `alloc_shape` -> `declare_shape` instead? It is only necessary for stack array objects whose shape is non-constant. This is important notably for all the f2py files

Reply:
So in your first example we have the following?
```
a.alloc_shape = (n, )
a.shape = (shape(a, 0), )

b.alloc_shape = (shape(a, 0), )
b.shape = (shape(b, 0), )
```
And you need `alloc_shape` for declaring the automatic Fortran arrays?

I am a little worried because such a declaration will be moved in the Fortran code to the declarations' section, so how can we ensure that the code is correct?

Reply:
Yes exactly.
`alloc_shape` is needed for declaring stack arrays. In the case of a stack array (all f2py arrays) the declaration has the following form:
```
subroutine f(n)

  implicit none
  integer(kind=8), value  :: n
  real(kind=8)  :: a (0:n-1)
```
Thus the declaration (which has no knowledge of the expression used to initialise `a`) needs to have the information about `n`.

I agree that this is dangerous. I am not sure that we could declare `b` as a stack array by writing it this way. However the previous way led to b having the correct size only if it was a stack array.
I tried to avoid needing to create the `alloc_shape` variable but `stack_array`s were the one case where I couldn't work out how to.
The only other thing that I can think of is to create a `ConstantVariable` if a Variable is used to declare the size of a stack_array. This way we could be sure that it does not change afterwards and then we can feel free to use this in the shape.

Maybe we should discuss this this afternoon

Reply:
Sounds good. Further improvements do not belong to this PR anyway.